### ISSUE TITLE
A possible bug in HistogramObserver._combine_histograms()

### ISSUE THREAD

**Issue by sterngerlach** [pytorch/pytorch#156414]
### 🐛 Describe the bug

Hello, I have a question about `_combine_histograms()` in `torch.ao.quantization.HistogramObserver`.
Maybe I'm misunderstanding the code, but I think the `bin_value` needs to be `torch.sum(orig_hist)` instead of `torch.sum(update_hist)`.
Does anyone know why it uses `update_hist` to count the number of samples in an original histogram?

https://github.com/pytorch/pytorch/blob/a8fe982993221048ee1665ce28add1b02888784d/torch/ao/quantization/observer.py#L1241-L1249

Also, it seems `_calculate_qparams()` in `UniformQuantizationObserverBase` uses `min_val_neg` and `max_val_pos` to compute quantization parameters (scale and zero-point) in case of asymmetric quantization (instead of the original `min_val` and `max_val`).
I would like to know the reason for this.

https://github.com/pytorch/pytorch/blob/a8fe982993221048ee1665ce28add1b02888784d/torch/ao/quantization/observer.py#L401-L405

Thank you!

### Versions

PyTorch version: 2.7.1+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.2 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: version 4.0.2
Libc version: glibc-2.39

Python version: 3.13.5 (main, Jun 19 2025, 07:55:24) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-60-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: 12.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @leslie-fang-intel @msaroufim

**Comment by Akabbaj** [comment]
@sterngerlach I do believe thats a bug in the implementation of _combine_histograms() and I've submitted a PR to fix it. 

As for your second question, I looked into it and while I’m not entirely certain, I think the key distinction lies in the ability to quantize and dequantize the real value 0.0  which appears often in neural networks. For example in ReLU activation functions or padding. 

For example, in the elif statement: [torch.per_channel_affine_float_qparams](https://github.com/pytorch/pytorch/blob/a8fe982993221048ee1665ce28add1b02888784d/torch/ao/quantization/observer.py#L395-L400), float quantization is used. Because the scale and zero point are floats, there’s always a 1-to-1 mapping between real values and quantized values in this case, so 0.0 is exactly representable.


However, in the else branch you linked, integer quantization is performed. This requires the zero point to be an integer in the quantization range. If 0.0 is not within [min_val, max_val], then sometimes  it can’t be quantized correctly and gets clamped. 

For example, if min_val = 2.0, max_val = 5.0, quant_min = 0, quant_max = 255, then if i'm not mistaken, the code snippet you shared [here](https://github.com/pytorch/pytorch/blob/a8fe982993221048ee1665ce28add1b02888784d/torch/ao/quantization/observer.py#L401-L405): would quantize 0.0 to 0 and then dequantize it to 2.0. 
 
Maybe someone with better understanding can shed more light on it!
